core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java

```diff
@@ -236,13 +256,41 @@
      * @param alloc the {@link ByteBufAllocator} which will allocate the content buffer
      * @param bufferSize the maximum allowed size of the {@link HttpData} buffers
      */
-    static StreamMessage<HttpData> of(Path path, ExecutorService executor, ByteBufAllocator alloc,
+    static StreamMessage<HttpData> of(Path path, @Nullable ExecutorService executor, ByteBufAllocator alloc,
                                       int bufferSize) {
         requireNonNull(path, "path");
-        requireNonNull(executor, "executor");
         requireNonNull(alloc, "alloc");
         checkArgument(bufferSize > 0, "bufferSize: %s (expected: > 0)", bufferSize);
-        return new PathStreamMessage(path, alloc, executor, bufferSize);
+        return of(path, executor, alloc, bufferSize, 0, Long.MAX_VALUE);
+    }
+
+    /**
+     * Creates a new {@link StreamMessage} that streams the specified {@link Path}.
+     * The specified {@code bufferSize} is used to create a buffer used to read data from the {@link Path}.
+     * Therefore, the returned {@link StreamMessage} will emit {@link HttpData}s chunked to
+     * size less than or equal to {@code bufferSize}.
+     *
+     * @param path the path of the file.
+     * @param executor the {@link ExecutorService} which performs blocking IO read.
+     * @param alloc the {@link ByteBufAllocator} which will allocate the content buffer.
+     * @param bufferSize the maximum allowed size of the {@link HttpData} buffers.
+     * @param start the start position of the file to read from; must be non-negative.
+     *              {@code 0} indicates the start of the file.
+     * @param end the end position of the file to read until, excluding the byte at the end position.
+     *            {@value Long#MAX_VALUE} indicates the end of the file.
+     */
+    @UnstableApi
+    static StreamMessage<HttpData> of(Path path, @Nullable ExecutorService executor, ByteBufAllocator alloc,
```

Comment:
At second thought, should we use a builder pattern instead?

```java
StreamMessage<HttpData> stream =
  StreamMessage
    .builder(path /* or file */) // returns `PathStreamMessageBuilder`
    .executor(...)
    .alloc(...)
    .range(start, end)
    .build();
```

Reply:
I am not sure it is good to add a dedicated builder method for `Path` because `StreamMessage` is a general data type that is used for any type.

Had a chat with @trustin and we decided:
- Add `range(...)` option to `HttpFile`
- Add `readStream()` for converting a `HttpFile` to a `StreamMessage`.
  ```scala
  val foo = httpFile.readStream(...)
  // Can access a metadata of a file from the returned value
  val metadataFuture = foo.metadata
  val contentStream = foo.content
  ```

Reply:
`HttpFile` supports not only a filesystem file using `Path` but also a classpath file using `InputStream`.
A general way to slice a `StreamMessage<HttpData>` would be needed.
Had a chat with other maintainers and we decided to add `ByteStreamMessage` that offers useful operations to handle byte streams such as `skipBytes()`, `takeBytes()` and `bufferSize()`.
As this PR came to a big change, I will add `range()` method to `HttpFile` in the following PR.
This PR will focus on `ByteStreamMessage` and its implementations.